readthedocs/search/api/v3/backend.py

```diff
@@ -0,0 +1,225 @@
+from functools import cached_property
+from itertools import islice
+
+from readthedocs.builds.models import Version
+from readthedocs.projects.models import Project
+from readthedocs.search.api.v3.queryparser import SearchQueryParser
+from readthedocs.search.faceted_search import PageSearch
+
+
+class Backend:
+
+    """
+    Parse the query, search, and return the projects used in the search.
+
+    :param arguments_required: If `True` and the user didn't provide
+     any arguments in the query, we don't perform the search.
+    :param default_all: If `True` and `arguments_required` is `False`
+     we search all projects by default, otherwise we search all projects
+     the user has access to.
+    :param max_projects: The maximum number of projects used in the search.
+     This limit is only applied for projects given explicitly,
+     not when we default to search all projects.
+    """
+
+    def __init__(
+        self,
+        *,
+        request,
+        query,
+        arguments_required=True,
+        default_all=False,
+        max_projects=100
+    ):
+        self.request = request
+        self.query = query
+        self.arguments_required = arguments_required
+        self.default_all = default_all
+        self.max_projects = max_projects
+
+    @cached_property
+    def projects(self):
+        """
+        Return all projects used in this search.
+
+        If empty, it will search all projects.
+
+        :returns: A list of tuples (project, version).
+        """
+        projects = islice(self._get_projects_to_search(), self.max_projects)
+        # Make sure we are using just one version per-project,
+        # searching multiple versions of the same projects isn't supported yet.
+        projects_dict = dict(projects)
+        return list(projects_dict.items())
+
+    def search(self, **kwargs):
+        """
+        Perform the search.
+
+        :param kwargs: All kwargs are passed to the `PageSearch` constructor.
+        """
+        if not self._has_arguments and self.arguments_required:
+            return None
+
+        projects = {project.slug: version.slug for project, version in self.projects}
+        # If the search is done without projects, ES will search on all projects.
+        # If we don't have projects and the user provided arguments,
+        # it means we don't have anything to search on (no results).
+        # Or if we don't have projects and we don't allow searching all,
+        # we also just return.
+        if not projects and (self._has_arguments or not self.default_all):
+            return None
+
+        search = PageSearch(
+            query=self.parser.query,
+            projects=projects,
+            **kwargs,
+        )
+        return search
+
+    def _get_projects_to_search(self):
+        if not self._has_arguments:
+            if self.arguments_required:
+                return None
+            yield from self._get_default_projects()
```

Comment:
It feels odd that this code is using `yield` where we aren't anywhere else. Is there a functionality reason for it? If so, we should clarify that somewhere in the file, because I'm not 100% comfortable reviewing code with it since we haven't used it much before.

Reply:
We are using yield/iterators at

https://github.com/readthedocs/readthedocs.org/blob/aa31543402a2c03e1f8a779e719266dc691b9671/readthedocs/search/parsers.py#L134-L139

It's easier than adding things to a list and then returning it (we have several sources from where to add things), it also helps us to put a limit at the number of projects we can search at the same time https://github.com/readthedocs/readthedocs.org/blob/adec99886dafd5fccba498bcab707b3a5cb0b40a/readthedocs/search/api/v3/backend.py#L49

otherwise we would have the list of all projects and then cut it, here we just stop at the `n`th project.

Reply:
This is useful context, we should write it in the code somewhere.

Reply:
Done!